src/app/OperationalSessionSetup.cpp

```diff
@@ -191,27 +191,29 @@ void OperationalSessionSetup::UpdateDeviceData(const Transport::PeerAddress & ad
     if (mState == State::ResolvingAddress)
     {
         MoveToState(State::HasAddress);
-        err = EstablishConnection();
-        if (err != CHIP_NO_ERROR)
-        {
-            DequeueConnectionCallbacks(err);
-            // Do not touch `this` instance anymore; it has been destroyed in DequeueConnectionCallbacks.
-            return;
-        }
-    }
-    else
-    {
-        if (!mSecureSession)
+        mInitParams.sessionManager->UpdateAllSessionsPeerAddress(mPeerId, addr);
+        if (!mPerformingAddressUpdate)
```

Comment:
I can't leave a comment in `Connect`, but what if the following sequence of things happened:

1. An exchange fails the first MRP, and we end up in `PerformAddressUpdate`. That sets `mPerformingAddressUpdate` to true and sets the state to `ResolvingAddress `. 
2. The exchange times out waiting for MRP _before_ the address resolution completes, and we mark the session defunct.
3. While we wait for address resolution to complete, someone else calls `FindOrEstablishSession`. They'll end up `Connect`, and find that we're in ResolveAddress state, and `isConnected` will be false, and return, expecting a success callback once CASE establishes. However, when we get to this line here, we're not going to continue to do CASE anymore, and their callback will fire because we call `DequeueConnectionCallbacks` below in line 209.

Reply:
The solution here seems simple - just clear `mPerformingAddressUpdate` in `Connect` I think.

Reply:
I think you missed one critical piece with `OperationalSessionSetup * FindExistingSessionSetup(const ScopedNodeId & peerId, bool forAddressUpdate = false)`

We have a completely separate `OperationalSessionSetup` just to do an address lookup, that is completely different from any `OperationalSessionSetup` for connecting.

See https://github.com/project-chip/connectedhomeip/pull/21768#discussion_r942959169 for more details

Reply:
Missed that, thanks!